### PR TITLE
Source Specific Multicast support

### PR DESCRIPTION
**Pull request by fatpelt**
Issue Number #332

### CODE REVIEW

**Comment by fatpelt** [comment]
i have validated it with a source yes.  we use SSM multicast pretty extensively at work. SSM is a pretty standard acronym, but i don't mind changing the references per your thoughts.  I thought about using the empty string, and don't mind. it wasn't immediately clear to me if there was a project standard on that.  also the spacing. 

i'll get all these fixed up and an updated pull request issued.  

i also realized that i need to update docs.  i'm assuming that docs/source/ is what's behind https://google.github.io/shaka-packager/html/tutorials/live.html ?

**Comment by kqyang** [comment]
Great. Thanks.

Yes, I prefer to spell out SSM, which makes it easier for people like me to read the code :). As for the empty string, I am fine with both options.

We do have a strict requirement on the coding styles, including alignment, spaces etc. We follow Chromium coding style: https://chromium.googlesource.com/chromium/src/+/master/styleguide/c++/c++.md. This helps code readability.

> i also realized that i need to update docs. i'm assuming that docs/source/ is what's behind https://google.github.io/shaka-packager/html/tutorials/live.html ?

Good catch. Yes, these are the two files you may want to touch: docs/source/options/udp_file_options.rst and docs/source/tutorials/live.rst.

**Comment by fatpelt** [comment]
ok.  i *think* i got them all.  the code i modified shows clean in the format check, but i'm betting i missed something.  also, sorry for the multiple commits rather than a rollback.  i'm still figuring this git thing out (i'm a cvs/svn old hat)

File: docs/source/options/udp_file_options.rst [hunk 1]
```diff
@@ -8,11 +8,16 @@ options:
...
-:interface=<addr>, source=<addr>:
+:interface=<addr>:
 
-    Multicast group interface address. Only the packets sent to this address is
+    Multicast group interface address. Only the packets sent to this address are
     received. Default to "0.0.0.0" if not specified.
 
+:source=<addr>:
+
+    Multicast source ip address.  Only the packets sent from this source address
```
Comment (kqyang) [inline docs/source/options/udp_file_options.rst]: Remove an extra space between sentences. We usually have only one space between sentences. similar below.

File: packager/file/udp_file.cc [hunk 2]
```diff
@@ -204,22 +204,54 @@ bool UdpFile::Open() {
   }
 
   if (is_multicast) {
-    struct ip_mreq multicast_group;
-    multicast_group.imr_multiaddr = local_in_addr;
+    if (options->is_source_specific_multicast()) {
+      // user has specified they want a specific source sender.  let's use that
+      // here
```
Comment (kqyang) [inline packager/file/udp_file.cc]: remove the comment. not useful here.

**Comment by kqyang** [comment]
Don't worry about the multiple commits. Github provides a handy function to squash and merge the changes into one.

**Pull request merged**